aten/src/ATen/cpu/vec256/missing_vld1_neon.h

```diff
@@ -0,0 +1,452 @@
+/* Workaround for missing vld1_*_x2 and vst1_*_x2 intrinsics in gcc-7.  */
+
+__extension__ extern __inline uint8x8x2_t
+__attribute__ ((__always_inline__, __gnu_inline__, __artificial__))
+vld1_u8_x2 (const uint8_t *__a)
+{
+  uint8x8x2_t ret;
+  asm ("ld1 {%S0.8b - %T0.8b}, [%1]" : "=w" (ret) : "r"(__a) :);
+  return ret;
+}
+
+__extension__ extern __inline int8x8x2_t
+__attribute__ ((__always_inline__, __gnu_inline__, __artificial__))
+vld1_s8_x2 (const int8_t *__a)
+{
+  int8x8x2_t ret;
+  asm ("ld1 {%S0.8b - %T0.8b}, [%1]" : "=w" (ret) : "r"(__a) :);
+  return ret;
+}
+
+__extension__ extern __inline uint16x4x2_t
+__attribute__ ((__always_inline__, __gnu_inline__, __artificial__))
+vld1_u16_x2 (const uint16_t *__a)
+{
+  uint16x4x2_t ret;
+  asm ("ld1 {%S0.4h - %T0.4h}, [%1]" : "=w" (ret) : "r"(__a) :);
+  return ret;
+}
+
+__extension__ extern __inline int16x4x2_t
+__attribute__ ((__always_inline__, __gnu_inline__, __artificial__))
+vld1_s16_x2 (const int16_t *__a)
+{
+  int16x4x2_t ret;
+  asm ("ld1 {%S0.4h - %T0.4h}, [%1]" : "=w" (ret) : "r"(__a) :);
+  return ret;
+}
+
+__extension__ extern __inline uint32x2x2_t
+__attribute__ ((__always_inline__, __gnu_inline__, __artificial__))
+vld1_u32_x2 (const uint32_t *__a)
+{
+  uint32x2x2_t ret;
+  asm ("ld1 {%S0.2s - %T0.2s}, [%1]" : "=w" (ret) : "r"(__a) :);
+  return ret;
+}
+
+__extension__ extern __inline int32x2x2_t
+__attribute__ ((__always_inline__, __gnu_inline__, __artificial__))
+vld1_s32_x2 (const int32_t *__a)
+{
+  int32x2x2_t ret;
+  asm ("ld1 {%S0.2s - %T0.2s}, [%1]" : "=w" (ret) : "r"(__a) :);
+  return ret;
+}
+
+__extension__ extern __inline uint64x1x2_t
+__attribute__ ((__always_inline__, __gnu_inline__, __artificial__))
+vld1_u64_x2 (const uint64_t *__a)
+{
+  uint64x1x2_t ret;
+  asm ("ld1 {%S0.1d - %T0.1d}, [%1]" : "=w" (ret) : "r"(__a) :);
```

Comment:
Also can help me understand this syntax a little bit or give some reference? Specifically what does '%S' and '%T' do? Thanks!

Reply:
%S0 and %T0 is gcc's way to say consecutive vector registers.  If you just say %w it will allocate vector registers without the extra constraint, and gcc will happily generate illegal vector ranges like "ld1 v7.1d - v3.1d" which is not allowed and will error in the assembler.

The same syntax is used in the machine description to specify the ld1 pattern:
https://patches.linaro.org/patch/118101/
```
+(define_insn "aarch64_simd_ld1<mode>_x2"
+  [(set (match_operand:OI 0 "register_operand" "=w")
+	(unspec:OI [(match_operand:OI 1 "aarch64_simd_struct_operand" "Utv")
+		    (unspec:VQ [(const_int 0)] UNSPEC_VSTRUCTDUMMY)]
+		   UNSPEC_LD1))]
+  "TARGET_SIMD"
+  "ld1\\t{%S0.<Vtype> - %T0.<Vtype>}, %1"
+  [(set_attr "type" "neon_load1_2reg<q>")]
+)
```
I back-ported to gcc-7 the patch I mentioned and two others (to add st1_x2, ld1_x3, etc.) and sent those patches to Amazon Linux 2 and Ubuntu 18.04 for integration in their gcc-7 compilers.
We will also add those back-ports to a gcc-7-arm-vendor branch for other distros to pick up.

Testing of the patch was minimal: I verified that pytorch compiles with gcc on Graviton2 aarch64-linux.
Could you please point me to how to run `make check` or a testsuite?  Thanks!

Reply:
Thanks for the explanation. Actually this is "kind" of what I was looking for when trying to introduce vec256 for arm32 via inline asm. Clang used to seg fault when using intrinsics in arm32 until recently: 
To work with consecutive registers of a specific name I used register keyword to get it to work in this PR: https://github.com/pytorch/pytorch/pull/41267/files#diff-fa20f55b1104b19819e148cc99e3482dR936. The reason for doing this was that otherwise I will have to do moves inside the asm block which will have perf implications. That PR has since been reverted for different reasons, but I would love to know if there is a better way of doing that.

For testing:
If you build pytorch for arm64 via something like
`BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 BUILD_CAFFE2_MOBILE=1 ANDROID_ABI=arm64-v8a ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON`
It should build `vec256_test` in build_android/bin directory. You can run that on an actual device.
Test is at: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/vec256_test.cpp